docs: add Identities documentation (#479)
canonical/pebble: docs/explanation/api-and-clients.md

# API and clients

The Pebble daemon exposes an API (HTTP over a unix socket) to allow remote clients to interact with the daemon. It can start and stop services, add configuration layers to the plan, and so on.

> If `pebble run` is started with the `--http <address>` option, Pebble also allows access to "untrusted" HTTP endpoints using the given TCP address. See "access control and identities" below.

Reviewer: ```suggestion
If `pebble run` is started with the `--http <address>` option, Pebble also allows access to "untrusted" HTTP endpoints using the given TCP address. See {ref}`explanation-api-access-control` below.
```

Author: Done.

We try to never change the underlying HTTP API in a backwards-incompatible way, however, in rare cases we may change the Go client in a backwards-incompatible way.

In addition to the Go client, there's also a [Python client](https://github.com/canonical/operator/blob/master/ops/pebble.py) for the Pebble API that's part of the [`ops` library](https://github.com/canonical/operator) used by Juju charms ([documentation here](https://juju.is/docs/sdk/interact-with-pebble)).

> 

Reviewer: ```suggestion
(explanation-api-access-control)=
```

Suggest adding an anchor here.

Author: Done.

## Access control and identities

API endpoints fall into one of three access levels, from least restricted to most restricted:

* `untrusted`: these are allowed from any user, even unauthenticated users using the HTTP-over-TCP listener. The only untrusted endpoints are `/v1/system-info` and `/v1/health`.

> * `read`: these are allowed from any authenticated user, regardless of access levey. They are usually read operations that use the HTTP `GET` method, such as listing services or viewing notices.

Reviewer 2: ```suggestion
* `read`: these are allowed from any authenticated user, regardless of access level. They are usually read operations that use the HTTP `GET` method, such as listing services or viewing notices.
```

Author: Fixed, thanks!

* `admin`: these are only allowed from admin users. They are usually write or modify operations that use the HTTP `POST` method, such as adding a layer or starting a service.

Pebble authenticates clients that connect to the socket API using peer credentials ([`SO_PEERCRED`](https://man7.org/linux/man-pages/man7/socket.7.html)) to determine the user ID (UID) of the connecting process. If this UID is 0 (root) or the UID of the Pebble daemon, the user's access level is `admin`, otherwise the access level is `read`.

If Pebble can't authenticate the user at all, or if it's an unauthenticated user connecting over TCP, the user is considered `untrusted`.

In addition to the default access control, Pebble admins can also set up named [identities](../reference/identities.md) with specific access levels using the identities CLI commands (see [how to manage identities](../how-to/manage-identities.md)).

> Currently the only identity type is `local`, that is, a local user ID determined using peer credentials. An example admin identity named "bob" is shown below:

Reviewer 2: Here I find the introduction of identity "type" a little abrupt, it was not introduced before, and then there is a sudden mention of identity, followed by an example.

I don't know how to make this better, but maybe move the sentence "Each identity has a name, an access level, and an authentication type with type-specific configuration." from the how-to doc here? Like:

```yaml
Each identity has a name, an access level, and an authentication type with type-specific configuration. An example of an identity named "bob" with admin access level and local type is shown below:

...

Currently the only identity type is `local`, that is, a local user ID determined using peer credentials.
```

Or, maybe we can have two subsections under `## Access control and identities`, one is `### Access level` and the other is `### Identity type`. Although this might be an overkill since the only identity type is local...

Reviewer: I agree the flow could be improved here.

I suggest something like this (this isn't perfect, open to suggestions).

## Access control and identities -> ## API access levels

API endpoints fall into...

If Pebble can't authenticate the user...

### Control API access by identity

In addition to...

Reviewer: And I would move the link to the "how to manage identities" towards the end of the section instead of early on.

Author: I've made these changes, thanks.
